dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/AnalyzerExecutable.java

```diff
@@ -0,0 +1,381 @@
+        }
+
+        // Figure out Dart version and use machine output when >=2.12
+        final String command = System.getProperty("os.name").toUpperCase().contains("WINDOWS")
+                ? "dart.bat"
+                : "dart";
+        final ProcResult result = new ProcBuilder(command, "--version")
+                .withWorkingDirectory(sensorContext.fileSystem().baseDir())
+                .run();
+
+        final int exitCode = result.getExitValue();
+
+        if (exitCode != 0) {
+            LOGGER.warn("Command '{}' finished (exit {})", result.getProcString(), exitCode);
+        } else {
+            LOGGER.debug("Command '{}' finished (exit {})", result.getProcString(), exitCode);
+        }
+
+        final String output = new String(result.getErrorBytes(), StandardCharsets.UTF_8).split("\\R", 2)[0];
+
+        // The version always written to stderr - https://github.com/dart-lang/sdk/issues/19704
+        // Need to use exit code to determine success
+        if (exitCode != 0) {
+            LOGGER.warn("Could not determine Dart version from output: '{}' - defaulting to legacy mode", output);
+            return AnalyzerOutput.Mode.LEGACY;
+        }
+
+        final Matcher matcher = Pattern.compile("^Dart SDK version: (.*) \\((.*)\\) \\((.*)\\) on (.*)$").matcher(output);
+        if (!matcher.matches()) {
+            LOGGER.warn("Could not determine Dart version from output: '{}' - defaulting to legacy mode", output);
+            return AnalyzerOutput.Mode.LEGACY;
+        }
+        final String version = matcher.group(1);
+        try {
+            final Semver semver = new Semver(version);
+            boolean isDart212 = semver.isGreaterThanOrEqualTo("2.12.0");
+            if (isDart212) {
+                LOGGER.info("Detected Dart SDK 2.12.0+ - using MACHINE mode for analyzer output.");
+                return AnalyzerOutput.Mode.MACHINE;
+            }
+            LOGGER.info("Detected Dart SDK <2.12.0 - using LEGACY mode for analyzer output.");
+            return AnalyzerOutput.Mode.LEGACY;
+        } catch (Exception e) {
+            LOGGER.warn("Could not detect Dart version from output: '{}' - defaulting to legacy mode", output, e);
+        }
+        return AnalyzerOutput.Mode.LEGACY;
+    }
+
+    private static AnalyzerExecutable.Mode getAnalyzerMode(SensorContext sensorContext) {
+        return sensorContext.config()
+                .get(DartAnalyzerSensor.ANALYZER_MODE)
+                .map(AnalyzerExecutable.Mode::valueOf)
+                .orElse(AnalyzerExecutable.Mode.defaultMode);
+    }
+
+    private static AnalyzerOutput.Mode getAnalyzerOutputMode(SensorContext sensorContext) {
+        return sensorContext.config()
+                .get(DartAnalyzerSensor.ANALYZER_OUTPUT_MODE)
+                .map(AnalyzerOutput.Mode::valueOf)
+                .orElse(AnalyzerOutput.Mode.defaultMode);
+    }
+}
+
+class DartanalyzerAnalyzerExecutable extends AnalyzerExecutable {
```

Comment:
The source file is quite big already, why not moving AnalyzerExecutable implementation to their own source file ?

Reply:
I agree, will do that tomorrow.